src/internal/jacobian.jl

```diff
@@ -0,0 +1,191 @@
+                          (linsolve === nothing || __needs_concrete_A(linsolve)))
+    alg_wants_jac = concrete_jac(alg) !== nothing && concrete_jac(alg)
+    needs_jac = linsolve_needs_jac || alg_wants_jac
+
+    @bb fu = similar(fu_)
+
+    if !has_analytic_jac && needs_jac
+        sd = __sparsity_detection_alg(f, autodiff)
+        jac_cache = iip ? sparse_jacobian_cache(autodiff, sd, uf, fu, u) :
+                    sparse_jacobian_cache(autodiff, sd, uf, __maybe_mutable(u, autodiff);
+            fx = fu)
+    else
+        jac_cache = nothing
+    end
+
+    J = if !needs_jac
+        JacobianOperator(prob, fu, u; jvp_autodiff, vjp_autodiff)
+    else
+        if has_analytic_jac
+            f.jac_prototype === nothing ? undefmatrix(u) : f.jac_prototype
+        elseif f.jac_prototype === nothing
+            init_jacobian(jac_cache; preserve_immutable = Val(true))
+        else
+            f.jac_prototype
+        end
+    end
+
+    return JacobianCache{iip}(J, f, uf, fu, u, p, jac_cache, alg, 0, autodiff, vjp_autodiff,
+        jvp_autodiff)
+end
+
+function JacobianCache(prob, alg, f::F, ::Number, u::Number, p; kwargs...) where {F}
+    uf = JacobianWrapper{false}(f, p)
+    return JacobianCache{false}(u, f, uf, u, u, p, nothing, alg, 0, nothing, nothing,
+        nothing)
+end
+
+@inline (cache::JacobianCache)(u = cache.u) = cache(cache.J, u, cache.p)
+@inline function (cache::JacobianCache)(::Nothing)
+    J = cache.J
+    J isa JacobianOperator && return StatefulJacobianOperator(J, cache.u, cache.p)
+    return J
+end
+
+function (cache::JacobianCache)(J::JacobianOperator, u, p = cache.p)
+    return StatefulJacobianOperator(J, u, p)
+end
+function (cache::JacobianCache)(::Number, u, p = cache.p) # Scalar
+    cache.njacs += 1
+    J = last(__value_derivative(cache.uf, u))
+    return J
+end
+# Compute the Jacobian
+function (cache::JacobianCache{iip})(J::Union{AbstractMatrix, Nothing}, u,
+        p = cache.p) where {iip}
+    cache.njacs += 1
+    if iip
+        if has_jac(cache.f)
+            cache.f.jac(J, u, p)
+        else
+            sparse_jacobian!(J, cache.autodiff, cache.jac_cache, cache.uf, cache.fu, u)
+        end
+        J_ = J
+    else
+        J_ = if has_jac(cache.f)
+            cache.f.jac(u, p)
+        elseif __can_setindex(typeof(J))
+            sparse_jacobian!(J, cache.autodiff, cache.jac_cache, cache.uf, u)
+            J
+        else
+            sparse_jacobian(cache.autodiff, cache.jac_cache, cache.uf, u)
+        end
+    end
+    return J_
+end
+
+# Sparsity Detection Choices
+@inline __sparsity_detection_alg(_, _) = NoSparsityDetection()
+@inline function __sparsity_detection_alg(f::NonlinearFunction, ad::AbstractSparseADType)
+    if f.sparsity === nothing
+        if f.jac_prototype === nothing
+            if is_extension_loaded(Val(:Symbolics))
+                return SymbolicsSparsityDetection()
+            else
+                return ApproximateJacobianSparsity()
```

Comment:
As a default?

Reply:
Symbolics is actually pretty good after a size threshold, and even faster than the approximate version.

Reply:
I mean defaulting to approximate Jacobian sparsity seems a bit odd. Even if it's not AD?